Services filtering and other stuff

marusak opened the pull request:
This PR implements #10010 and #11237.
This PR is WIP and I am happy to improve further if design moves (especially in #11237).
I am opening as there are a lot of questions for design that rose as I was implementing this. (See screenshots at the end, there are the issues visible). Mostly asking @andreasn as he is reporter of both issues, but every single option would be great!

1. How should the page look like when there are no matches for current filters? (I put there just something very fast, I don't like it that the page can be blank)
2. Under Timers there is another button, should it be on far right or left? (Makes sense to have on left, but rather asking)
3. On filtering columns can resize and they often do as for example you filter our some service with very long name. Is this okay, or not? If not, what can be done?
4. Label `Automatic Startup` is rather long. On PC screen it is fine, problem is on small screens. If I put there wrapping then it wraps on PC screen (surely can be limited by screen size). Would for example `Autostart` be suitable?
5. Label `All services` or some state for empty search (sth. like `No services...`) can be confusing as there are button `Targets`, `Sockets`, `Timers`, `Paths` and you would filter them by `All services`. And also it is filtering autostart options.
6. I am confused a bit by `Templates` behavior (very likely I misunderstood templates) -  but now they can be under all Enabled/Disabled/Static. Would it make sense to make it own "class"?

Screenshots:
![emptystate](https://user-images.githubusercontent.com/12330670/53301832-d6e2aa80-3857-11e9-9347-47bf74144693.png)
![timers_automtic_startup](https://user-images.githubusercontent.com/12330670/53301833-d77b4100-3857-11e9-9dfa-ca23654a8d76.png)
![disbaled_services](https://user-images.githubusercontent.com/12330670/53301834-d77b4100-3857-11e9-8ac0-0dfb5ceec522.png)
![overview_services](https://user-images.githubusercontent.com/12330670/53301835-d77b4100-3857-11e9-964b-a715d927ec30.png)

TODOs:

- [x] Use listing
- [x] Empty page
- [x] Wrapping of `Automatic Startup` on small screens
- [x] Change `All Services` by page
- [x] Filter placeholder
- [x] Column widths
- [x] Case insensitive search
- [x] Dropdown menu with pull-left
- [x] Design approval
- [x] Open issue for templates
- [x] Tests

andreasn commented:
> 1. How should the page look like when there are no matches for current filters? (I put there just something very fast, I don't like it that the page can be blank)

Good question. I didn't consider this!
The [Patternfly design guidelines on filters](https://www.patternfly.org/pattern-library/forms-and-controls/filter/#design) suggests something like this:
![mockup](https://raw.githubusercontent.com/cockpit-project/cockpit-design/5b4a550e971708a67d28ccf57f8c421ddc3803c7/services/services-listing.png)

andreasn commented:
>     4\. Label `Automatic Startup` is rather long. On PC screen it is fine, problem is on small screens. If I put there wrapping then it wraps on PC screen (surely can be limited by screen size). Would for example `Autostart` be suitable?

I went for _Automatic Startup_ since it matches what we show on the Services details page where it forms the sentence "Automatic Startup Enabled". It also feels like more natural language to me.
I think it's OK if the header label breaks into two lines on smaller screens.

andreasn commented:
> Label `All services` or some state for empty search (sth. like `No services...`) can be confusing as there are button `Targets`, `Sockets`, `Timers`, `Paths` and you would filter them by `All services`. And also it is filtering autostart options.

Do you think it could work to relabel the dropdown "All Timers" etc. on the Timers page (and same for all the other pages). I realize this ends up being a bunch more work, but it would reflect the notion that the filter only applies to the current page.

andreasn commented:
> I am confused a bit by `Templates` behavior (very likely I misunderstood templates) - but now they can be under all Enabled/Disabled/Static. Would it make sense to make it own "class"?

I honestly don't understand how systemd templates work. Maybe @martinpitt can shed some light.

martinpitt commented:
Yeah, maybe templates should have their own class. You can enable/disable/start/stop a template, which will create instances -- But you always have to give it an actual instantiation value, so this requires a dialog.

I. e. it makes no sense to start/stop `getty@.service`, as the template needs an actual terminal device. You can start/enable two instances like `getty@tty1` and `getty@ttyS0`, and then stop/disable these.

Please see "man systemd.unit" for details about templates, but in short it's a way to build a family of similar services which differ only in a parameter (usually a device name or something such).

marusak commented:
Thank you @andreasn for answers, I will incorporate them.
Also thank you @martinpitt for explaining templates. I understood them in the same way, still does not explain though why should `getty@.service` be now under enabled , `httpd@.service` under disabled and `sshd@.service` in static. Anyway, should it become 4th option in the dropdown menu? And also what should be written in column `State` for templates? (`Template`, nothing?) @andreasn

martinpitt commented:
> still does not explain though why should getty@.service be now under enabled , httpd@.service under disabled and sshd@.service in static

Right, that does make no sense. Templates by themselves are nothing by themselves, we should only show the instances. These *can* be enabled or static.

martinpitt commented:
Sorry, I meant "show the instances in their respective categories". Templates themselves can/should of course still be shown, but in their separate category.

andreasn commented:
Does it currently only filter based on Id's? If so, it's probably best to mention that in the filter inputs placeholder text to give a hint of how the search will function.
What about _Filter by id_ rather than _Type to filter..._

andreasn commented:
>     3\. On filtering columns can resize and they often do as for example you filter our some service with very long name. Is this okay, or not? If not, what can be done?

I played around with this now. It jumps around a little bit, but it's not too bad.
The only solution I can think of to fixing the jumping would be to give the columns a fixed width as a percentage of the page, so that the service-unit-description always gets, say 35% of the screen estate.

marusak commented:
> Sorry, I meant "show the instances in their respective categories". Templates themselves can/should of course still be shown, but in their separate category.

A lot of confusion between category, class, type... :) 
Maybe would be great if @andreasn could draw design.  It would make sense to have two tables in each category (Timers, Services...) one for real instances and one for templates. But that would be step back. Or make another category `Templates` and put it next to `Paths`? But Templates can be for timers or services. Or put it into `autostart` (makes less sense, but could be the least confusing?)

> Does it currently only filter based on Id's?

Ids and description. But yes, design states also having there `Filter by id or description...` so I will go with it.

> The only solution I can think of to fixing the jumping would be to give the columns a fixed width as a percentage of the page, so that the service-unit-description always gets, say 35% of the screen estate.

That would be a solution. Kinda hard to guess the widths though :) But surely I can play with it and see if there is something reasonable that can be used. (Mind that timers need custom widths as there are more columns)

andreasn commented:
@garrett pointed out that the text for the no-results-found state was a little too complex, so changing that to _No Matching Results_
![mockup](https://raw.githubusercontent.com/cockpit-project/cockpit-design/1d1714403f93ab1910b84704f141ae2c27e36456/services/services-listing.png)

andreasn commented:
> On filtering columns can resize and they often do as for example you filter our some service with very long name. Is this okay, or not? If not, what can be done?

With the help of @garrett I figured out a way to make the columns not jump around so much.
Just drop this into pkg/systemd/services.css:
```
#services-list table {
    table-layout: fixed;
}

.service-unit-description {
    width: 40%;
}
```

It will of course still jump a little bit when the web browser adds and removes the scrollbar due to the page height, but that tolerable.

andreasn commented:
I just noticed that the text is centered here instead of left aligned, but I can't quite figure out why.
![screenshot from 2019-02-25 17-56-06](https://user-images.githubusercontent.com/209895/53354240-a74ea380-3926-11e9-895a-568a92ef4b71.png)

andreasn commented:
It probably makes most sense to spin off any changes to Templates into it's own issue. I don't think it's really blocking this PR.

andreasn commented:
> Ids and description. But yes, design states also having there `Filter by id or description...` so I will go with it.

Awesome. I just realized that the description seems case-sensitive. Is there a quick way to make it not case-sensitive? That would make searching for the description easier.

andreasn commented:
Seems the dropdown centers the text due to the `btn` class in Patternfly.
It is solved elsewhere with a `pull-left` class on the `span` that surrounds the text label. Not super-elegant, and @garrett is working on a new dropdown widget in https://github.com/cockpit-project/cockpit/pull/11155 that we can use instead, where that wouldn't be needed. Do we want to block on that PR landing first?

marusak commented:
> With the help of @garrett I figured out a way to make the columns not jump around so much.

Awesome! Thanks :)

> It probably makes most sense to spin off any changes to Templates into it's own issue. I don't think it's really blocking this PR.

Agree. I will open PR after this one gets merged (or at least design stabled).

> Is there a quick way to make it not case-sensitive?

Sure, adding into TODO.

> Not super-elegant, and @garrett is working on a new dropdown widget in #11155 that we can use instead, where that wouldn't be needed. Do we want to block on that PR landing first?

It is developed in React. Services page is not in React yet.

andreasn commented:
> It is developed in React. Services page is not in React yet.

Ah, ok. Let's use the pull-left class on the span then.

garrett commented:
> It will of course still jump a little bit when the web browser adds and removes the scrollbar due to the page height, but that tolerable.

It's possible to make the page not jump in that case, with `overflow-y: scroll` on the element that will scroll. (It's often, but not always, `body`.)

You could, for example, set `overflow-y: scroll;` on `body` when you're going to filter and remove it when you're not filtering. Or we could always set that attribute for the page, as it's usually a long table-based list of items.

garrett commented:
> Ah, ok. Let's use the pull-left class on the span then.

It would most likely be better to use `display: flex` (or `display: grid`, depending).

garrett commented on pkg/systemd/services.html:
```diff
@@ -56,13 +59,27 @@
         <button class="btn btn-default" role="tab" data-pattern="\.path$" translatable="yes">Paths</button>
       </div>
-      <button id="create-timer" data-toggle="#timer-dialog" data-target="#timer-dialog" class="btn btn-primary pull-right" translate>Create Timer</button>
+      <div>
```
There's an empty `<div>` here. Usually this comes up with React, where `<React.Fragment>` should be used. I'm not sure why there's an empty div here though. If you need to group elements in a wrapper, try to use a semantic element (when it makes sense) and please _at least_ give it a class and/or ID regardless to describe why this is here. (Something like "services-filter" might work.)

andreasn commented:
> As noted above, the dropdown has some issues, but we're hoping to solve those elsewhere (and backport it to the non-React parts of Cockpit, like here).

What do you think is the timeline for the non-react-dropdown fixes? Is it blocking this PR on fixing that first, or should we do the pull-left hack?

garrett commented:
> What do you think is the timeline for the non-react-dropdown fixes?

1. I need to include the CSS.
   - To do so, I need to include an SVG somehow. @larskarlitski and I are looking into this. @martinpitt might need to look into it too when he's back.
2. Then it needs to be tweaked from the CSS side of things and tested everywhere.

> Is it blocking this PR on fixing that first, or should we do the pull-left hack?

No, it's not blocking on the dropdown.

We should all avoid `pull-left` and using floats for layout (especially floats that change display order from visible content order). It's just bad code and a bad concept.

I don't think it has anything to do with the dropdown though? There are other ways to solve it?

garrett commented:
Oh, I looked at it, and `pull-left` is fine in this usage, as:
1. it doesn't change order of elements
2. it is already contained (so it wouldn't escape the parent element)

Go for it. :+1:

marusak commented:
Thank you guys for all the comments and ideas!!
I went through all of them and fixed them. I pushed new version. I will comment on few things, what I do not comment on is done.

> There's an empty \<div\> here. Usually this comes up with React, where <React.Fragment> should be used. I'm not sure why there's an empty div here though

It is to make it work with flex. I want some controls to be on left and some on right. So I split it into two divs, put there `justify-content: space-between`. I guess it can be done better. Anyway put there class to make it at least a bit more understandable.

> Hitting escape doesn't clear the filter box.

I either don't understand what you are trying to do, or cannot reproduce. I can write text into filter box and then delete it just fine with escape - it is standard `input` with no magic at all.

> Filter isn't focused by default (this is debatable, but probably would be nice feature-wise — there's a possible accessibility issue by skipping the toggle buttons on the left, however).

I don't have any strong options about this. Kinda it feels strange to focus search field (especially when there are other selectors left of them). But I am fine with implementing it if I am told so.

> The new dropdown widget is a drop-in replacement for React components, but copy/pasting the HTML (when it is done and merged) should be possible, as the CSS will be in the global CSS used for Cockpit.

Sounds well, but I would not do it in this PR. It should be separate PR that will do it on all non-react place of Cockpit.

Anyway, anything discussed here should be fixed. I also went through and used listing instead of what we currently had. It works nicely with small screens (current the table breaks very weirdly on small screens) buuut there are two isses - one is in design and now sure if can be fixed (possible reason why this wont work and other solution needs to be found) other is just styling.

First issue - between two last columns there is that big gap - it is because second to last column floats left and last to right. Is this a problem? What style can be used to improve this?
![newdesign](https://user-images.githubusercontent.com/12330670/53429117-49848f00-39ec-11e9-89bb-2acda3579e42.png)

Second issue - On smaller screens columns can break layout - notice how last items second column jumps to left.
![breaksmallscreen](https://user-images.githubusercontent.com/12330670/53429222-7a64c400-39ec-11e9-9c31-a4810df3f163.png)

marusak commented:
A bit of thinking about the second issue - the breaking layout. Possibly grid, where each column would be some percentage could work. We would need a few classes as different tables may have different number of columns, but surely could be done. Ideas?

garrett commented:
> I either don't understand what you are trying to do, or cannot reproduce. I can write text into filter box and then delete it just fine with escape - it is standard `input` with no magic at all.

I simply typed into the box and hit the ESC key. Nothing happened. It should clear the entry (but keep focus).

---

The layout fixing should be in another PR, hence me opening up https://github.com/cockpit-project/cockpit/issues/11258 earlier today. It's too complex to block this PR, and it is an existing problem. It's probably tricky with lots of HTML/CSS changes too. (The HTML changes would most likely be attribute related, not massive restructuring.)

More details about making accessible tables are in https://github.com/cockpit-project/cockpit/issues/11260. But I guess that's more for react tables and we'd handle services similarly, but sadly as a one-off, as it isn't React based.

garrett commented:
> Anyway put there class to make it at least a bit more understandable.

Thanks! As a rule, we should never, ever have `<div>` or `<span>` in our codebase. They should be there for a reason and have a class (for CSS) and/or ID (for JavaScript).
- Even if there's no CSS, it's good to have a name for it, as it helps to explain why it's there and makes it something to target in CSS in the future too.

- And an ID is also useful outside of JavaScript, for targeting the element in page (which we don't really do much in Cockpit).

marusak commented:
> I simply typed into the box and hit the ESC key. Nothing happened. It should clear the entry (but keep focus).

Ah, my mistake, I was thinking different key :smiley: Should this work? I never used ESC key to delete text I've inserted. (maybe it's the trick I am looking for, because often I just do ctrl+a and del)

> The layout fixing should be in another PR, hence me opening up #11258 earlier today.

Sure. I was not trying to implement the table to work properly on small devices or being accessible. I just implemented design that Andreas prepared. He suggested using listing. I can drop this commit easily, it just looks better :) 

>  As a rule, we should never, ever have \<div\> or \<span\> in our codebase.

Didn't know this, but makes sense :) Thanks for explaining this to me!

marusak commented:
> Should this work?

Yes, it should if the `input` has `type="search"`. I will update this. It works with chrome with this change, but seems that it does not work with firefox. https://bugzilla.mozilla.org/show_bug.cgi?id=1055085

andreasn commented:
Looks good to me now!
Only thing I noticed is that the cursor when you hover the cell is `cursor: default;` rather than `cursor: pointer;`
https://developer.mozilla.org/en-US/docs/Web/CSS/cursor#Values

andreasn commented:
Looks good to me now!

marusak commented:
Seems it is ready for code review.  @martinpitt can I ask you?
Current screenshot:
![currentservices](https://user-images.githubusercontent.com/12330670/53497644-eacf1c00-3aa4-11e9-8cf2-9d5d05cb6029.png)

garrett commented:
OK, I think it's good for a review from my end too.

garrett commented:
@marusak: Thanks again for all your work on this! :tada:

It's exciting to see filtering implemented, especially since people have wanted this in the usability studies... and I've wanted this exact feature on my own home server (running Cockpit) too. :wink:

garrett commented:
I _think_ the filter area gets weird when there's a create timer button, as it looks like it's part of the filter area. We'd want to add space between them.

It's also odd that there's [**filtering**] [**create timer**] [**more filtering**]

![screenshot_2019-02-27 services - sunny 1](https://user-images.githubusercontent.com/10246/53506549-78673780-3ab6-11e9-9b7c-216b0282aefa.png)

I'm OK with the fix being another PR.

marusak commented:
> We'd want to add space between them.
> It's also odd that there's [filtering] [create timer] [more filtering]

Space can be added. Also I am happy to adjust to any design (in this or another PR I dont mind). It was one of my questions when I opened PR where this button should be :)

marusak commented:
I don't want to overflow this PR with changes, but one more I think would be suitable :)
I did not push it yet, will push if agreed. It is fixing issue #698 . There are no functional changes just changing html. @andreasn @garrett ?
Screenshot:
![navbars](https://user-images.githubusercontent.com/12330670/53543813-37a60780-3b24-11e9-9727-b3bd5f8c28ae.png)

marusak commented:
@garrett I don't feel like commit `services: Add width to service-unit-id` is great solution.
Without this commit it looks like this:
![after](https://user-images.githubusercontent.com/12330670/53544103-34f7e200-3b25-11e9-9985-d0531c64e800.png)
And with the commit it looks like this:
![before](https://user-images.githubusercontent.com/12330670/53544117-4345fe00-3b25-11e9-981d-669d5ccce35a.png)
Which I think is much worse. I think it is totally fine to break the id on smaller screen, just not to break it unless needed.  I did something similar with label `Automatic startup` that it wraps only on screens with width smaller then 960px (see commit `services: Do not split into 3 tables, but use one `)

martinpitt commented on pkg/systemd/init.js:
```diff
@@ -325,6 +330,8 @@ $(function() {
             }
         }
 
+        $("#services-text-filter").on("input", render);
```
Haha, poor man's React :wink:

martinpitt commented on pkg/systemd/services.html:
```diff
@@ -56,7 +56,10 @@
-      <button id="create-timer" data-toggle="#timer-dialog" data-target="#timer-dialog" class="btn btn-primary pull-right" translate>Create Timer</button>
+      <div class="filter-group">
+        <button id="create-timer" data-toggle="#timer-dialog" data-target="#timer-dialog" class="btn btn-primary" translate>Create Timer</button>
+        <input type="search" id="services-text-filter" class="form-control" aria-label="Filter" placeholder="Filter by id or description..." translate="placeholder"/>
```
placeholder needs i18n. Also, shouldn't "id" be written as "ID"? 

systemd calls that "unit name". As we want to avoid jargon as much as possible, "Filter by name or description..." would be better. But as we already show "ID" in the table, let's postpone that part for another time.

marusak replied:
> Also, shouldn't "id" be written as "ID"?

In the table columns it is called `Id`. Design has `id`. 
But agree that `name` would be more suitable, but lets to this in another PR :)

marusak replied:
> placeholder needs i18n

`translate="placeholder"` is already there

martinpitt commented on test/verify/check-services:
```diff
@@ -204,6 +204,36 @@ Unit=test.service
+        init_filter_state()
+        b.set_input_text("#services-text-filter", "failing")
+        wait_service_not_present("test.service")
+        wait_service_present("test-fail.service")
```
I suggest adding another case that clears the input field again, to make sure that this works as well.

marusak replied:
`init_filter_state()` does exactly that - clear input and check that both services are present.

martinpitt replied:
Right, missed that. Thanks!

martinpitt commented on pkg/systemd/services.html:
```diff
@@ -171,6 +171,17 @@
+  <script id="service-empty-tmpl" type="x-template/mustache">
+    <div id="empty-search" class="blank-slate-pf">
+      <h1>
+        No Matching Results
```
Needs i18n

martinpitt commented on pkg/systemd/services.html:
```diff
@@ -171,6 +171,17 @@
+        No Matching Results
+      </h1>
+      <p>
+        <a id="clear-all-filters" href="#">Clear All Filters</a>
```
Needs i18n

martinpitt commented on pkg/systemd/init.js:
```diff
@@ -297,11 +297,15 @@ $(function() {
             });
```
This "Show message on empty result search" commit should be squashed into the previous one.

marusak replied:
Sure. But will squash into the following one, as it makes more sense that way.

martinpitt commented on pkg/systemd/init.js:
```diff
@@ -148,6 +148,14 @@ $(function() {
+        "\\.socket$": _("All Sockets"),
+        "\\.timer$": _("All Timers"),
+        "\\.path$": _("All Paths")
+    };
```
I don't quite understand this: we already have the tab headers for "Targets", "Services", "Sockets", and so on. Why do we need another selector for the type?

Also, this seems to have nothing to do with "filtering by autostart". Separate commit?

marusak replied:
Yeeah, I was expecting this will receive a comment. (I am not happy with it but cannot think of better way of doing this).
This is here, because when you change type for example from services to timers you want the dropdown contain "All Timers" and not "All Services" etc.
Was thinking how to do this so it still can be translated, can be dynamic and also I need to check if the current selection is not one of "All ..." and update that as well as  first item in the dropdown.

martinpitt replied:
Thanks, this makes more sense now. So conceptually this is "All enablement states", but that sounds too ugly. Perhaps "Any Startup", to relate it to the column header? But I'll leave this between @garrett and you, I just wanted to understand what's going on.

marusak replied:
It was suggested by @andreasn, so maybe he can join this as well

andreasn replied:
If you're concerned about repetition, _All_ could work as a label.

andreasn replied:
Or _Everything_

marusak replied:
Changed to `All`

martinpitt commented on pkg/systemd/services.css:
```diff
@@ -1,5 +1,5 @@
 #services-page {
-  verflow-y: scroll;
+    overflow-y: scroll;
```
There are two commits "Always show scroll bar". Squashing makes sense IMHO, as they read pretty well exactly the same.

marusak replied:
Of course, I did typo while rebasing and garrett caught it. But I thought he squashed it. Will squash it.

martinpitt commented on pkg/systemd/services.html:
```diff
@@ -19,27 +19,27 @@
           <thead>
             <tr>
-              <th scope="col">{{Id}}</th>
+              <th scope="col" class="service-unit-id">{{Id}}</th>
```
These aren't being used anywhere, in CSS nor tests. Is that planned?

marusak replied:
@garrett ?

garrett replied:
I had them in CSS at some point.

This specific one was being used for the width calculation. The CSS was targeting both the `th` and `td` for fixing the width... I looked and found in in the CSS locally, but updated the branch and don't see it anymore; I guess that changed at the end of last week?

garrett replied:
Also, I come across unclassed `<th>`s,` <td>`s all the time in Cockpit and see people target them with `:first-child`, `:last-child`, `:nth-child()` and such — we shouldn't target tables (or most elements) like this, unless it's really supposed to be the first or last element and not a specific row. That's why I added it.

(Although, we can fix it on the fly if we do need to target a specific element too.)

marusak commented:
Thank you @martinpitt for the review. I've dropped the last commit, squashed the commit with empty state and two with overflow and added two forgotten i18n.

> These aren't being used anywhere, in CSS nor tests. Is that planned?

This was introduced by @garrett so lets ask him.

Also I changed title so tests can run and I can see if there are some problems that need to be addressed.

marusak commented:
Aaa, I pushed one more commit I did not plan to push (moving from button group into tabs - see my comments earlier today). I keep it here for now, but I am ready to drop it if anyone is against. (I just think it will break tests)

martinpitt commented on pkg/systemd/services.html:
```diff
@@ -168,6 +171,17 @@
+  <script id="service-empty-tmpl" type="x-template/mustache">
+    <div id="empty-search" class="blank-slate-pf blank-screen">
+      <h1 translate>
+        No Matching Results
```
The whitiespacing doesn't work, it gets copied literally to cockpit.pot:

    msgid "        No Matching Results      "

Please write it as `<h1>Bla</h1>`.

marusak replied:
Ach, did not notice that. Thank you for pointing this out. Repushed

martinpitt commented on pkg/systemd/services.html:
```diff
@@ -62,6 +62,19 @@
       <div class="filter-group">
         <button id="create-timer" data-toggle="#timer-dialog" data-target="#timer-dialog" class="btn btn-primary" translate>Create Timer</button>
         <input type="search" id="services-text-filter" class="form-control" aria-label="Filter" placeholder="Filter by Id or description..." translate="placeholder"/>
+        <div class="dropdown" id="services-dropdown">
```
FTR, in the spirit of #11155 this should become an actual `<select>` soon. Right now the selector isn't keyboard navigatable at all. Fine for a follow-up of course.

garrett commented on pkg/systemd/services.html:
```diff
@@ -51,14 +51,14 @@
+        <li class="active" aria-current=true tabindex="0" data-pattern="\.service$"><a translatable="yes">System Services</a></li>
+        <li tabindex="0" data-pattern="\.socket$"><a translatable="yes">Sockets</a></li>
+        <li tabindex="0" data-pattern="\.timer$"><a translatable="yes">Timers</a><li>
+        <li tabindex="0" data-pattern="\.path$"><a translatable="yes">Paths</a><li>
```
Ah, cool to see tabindex as the tabs need to be selected — but why is it in `<li>` instead of `<a>`? (Better yet would be to add `href` to the `<a>`, instead of using a `tabindex`, and use [a URL fragment*](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#Attributes) to remember the state in the URL.)

_* URL fragment (`#this-thing-here`), not to be confused with a Document Fragment, React.Fragment, CSS Fragmentation, or anything else called "fragment" in webdev._ :wink:
